### ISSUE TITLE
Output gradient information in neural nets

### ISSUE THREAD

**Issue by lisitsyn** [shogun-toolbox/shogun#2723]
Currently debug output on neural net only reports the current training error. It is also worth to look at gradients information like mean gradient absolute value per layer

**Comment by gittyadi** [comment]
If i may, I'd like to have a go at this. Could you please point me to the code?

**Comment by lisitsyn** [comment]
@adityaosp95 feel free! The code reporting error can be found at https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/neuralnets/NeuralNetwork.cpp. See functions `train_bfgs` and `train_gradient_descent`.

I suggest you to use any neural network example to develop and test gradient output.

**Comment by sanuj** [comment]
@lisitsyn what information about the gradient has to be outputted in particular apart from `mean gradient absolute value per layer`?

**Comment by sanuj** [comment]
I tried running the examples of neural nets. I am not able to see the output of `SG_INFO` (though i can see the output of `std::cout`) in the terminal even after adding this:
```
void print_message(FILE* target, const char* str)
{
    fprintf(target, "%s", str);
}

void print_warning(FILE* target, const char* str)
{
    fprintf(target, "%s", str);
}

void print_error(FILE* target, const char* str)
{
    fprintf(target, "%s", str);
}
init_shogun(&print_message, &print_warning,
      &print_error);
```

I'm using `g++ -lshogun` to run the example and it's compiling fine. Gist of the code:
https://gist.github.com/sanuj/a9c5a4ec3fe2fb8af5b8

**Comment by lisitsyn** [comment]
@sanuj sorry for late answer
1) It's probably useful to check mean and max gradient values per layer
2) Use ```set_loglevel``` of ```SGIO```

**Comment by arasuarun** [comment]
Is this issue still open? #2840 seems to cover it... or is there anything else left to do?

**Comment by arasuarun** [comment]
@lisitsyn The entire output loop could probably be put into a separate function and vectorized, too. It's not exactly a pretty sight in the train_gd function as of now

What do you think?
